net/portmapper/upnp.go

```diff
@@ -130,40 +132,70 @@ func addAnyPortMapping(
 	return externalPort, err
 }
 
-// getUPnPClients gets a client for interfacing with UPnP, ignoring the underlying protocol for
+// getUPnPClient gets a client for interfacing with UPnP, ignoring the underlying protocol for
 // now.
 // Adapted from https://github.com/huin/goupnp/blob/master/GUIDE.md.
-func getUPnPClient(ctx context.Context, gw netaddr.IP) (upnpClient, error) {
+//
+// The gw is the detected gateway.
+//
+// The uPnPRootIPP is the most recently heard Location header's
+// ip:port from a UPnP discovery packet.
+func getUPnPClient(ctx context.Context, gw netaddr.IP, uPnPRootIPP netaddr.IPPort) (upnpClient, error) {
 	if controlknobs.DisableUPnP() {
 		return nil, nil
 	}
-	ctx, cancel := context.WithTimeout(ctx, 250*time.Millisecond)
-	defer cancel()
-	// Attempt to connect over the multiple available connection types concurrently,
-	// returning the fastest.
 
-	// TODO(jknodt): this url seems super brittle? maybe discovery is better but this is faster
-	u, err := url.Parse(fmt.Sprintf("http://%s:5000/rootDesc.xml", gw))
+	var port uint16
+	if uPnPRootIPP.IsZero() {
+		port = 5000
```

Comment:
Is there anyway to unset root IP if a network change is detected?

Reply:
Cleared in invalidateMappingsLocked now.